app/init/resources.php

```diff
@@ -781,6 +783,49 @@ function getDevice(string $root, string $connection = ''): Device
     return [];
 });
 
+App::setResource('devKey', function (Request $request, Document $project, array $servers, Database $dbForPlatform) {
+    $devKey = $request->getHeader('x-appwrite-dev-key', $request->getParam('devKey', ''));
+
+    // Check if given key match project's development keys
+    $key = $project->find('secret', $devKey, 'devKeys');
+    if (!$key) {
+        return new Document([]);
+    }
+
+    // check expiration
+    $expire = $key->getAttribute('expire');
+    if (!empty($expire) && $expire < DatabaseDateTime::formatTz(DatabaseDateTime::now())) {
+        return new Document([]);
+    }
+
+    // update access time
+    $accessedAt = $key->getAttribute('accessedAt', '');
+    if (DatabaseDateTime::formatTz(DatabaseDateTime::addSeconds(new \DateTime(), -APP_KEY_ACCESS)) > $accessedAt) {
+        $key->setAttribute('accessedAt', DatabaseDateTime::now());
+        Authorization::skip(fn () => $dbForPlatform->updateDocument('devKeys', $key->getId(), $key));
+        $dbForPlatform->purgeCachedDocument('projects', $project->getId());
+    }
+
+    // add sdk to key
+    $sdkValidator = new WhiteList($servers, true);
+    $sdk = $request->getHeader('x-sdk-name', 'UNKNOWN');
+
+    if ($sdkValidator->isValid($sdk)) {
```

Comment:
in case of `$sdk` being `UNKNOWN` from the fallback, we dont even need to use the validator. we should probably fallback to `null` on that header and skip this part without even setting up the validator.